File: rxfirebase2-config/src/main/java/rxfirebase2/config/RxFirebaseRemoteConfig.java
package rxfirebase2.config;

import com.google.firebase.remoteconfig.FirebaseRemoteConfig;

import io.reactivex.Completable;
import io.reactivex.annotations.CheckReturnValue;
import io.reactivex.annotations.NonNull;
import io.reactivex.functions.Action;
import rxtasks2.RxTask;

public final class RxFirebaseRemoteConfig {
    /**
     * @param config
     * @return
     */
    @CheckReturnValue
    @NonNull
    public static Completable fetches(@NonNull final FirebaseRemoteConfig config) {
        return RxTask.completes(config.fetch()).doOnComplete(new Action() {
            @Override
            public void run() throws Exception {
                config.activateFetched();
            }
        });
    }

    @CheckReturnValue
    @NonNull
    public static Completable fetches(@NonNull final FirebaseRemoteConfig config,
                                      final long timeout) {
        return RxTask.completes(config.fetch(timeout)).doOnComplete(new Action() {
            @Override
            public void run() throws Exception {
                config.activateFetched();
            }
        });
    }

    @CheckReturnValue
    @NonNull
    public static Completable fetchesDev(@NonNull final FirebaseRemoteConfig config) {
        return fetches(config, (config.getInfo().getConfigSettings().isDeveloperModeEnabled())
                ? 0L : 3600L);
    }
}

File: rxfirebase2-config-kotlin/src/main/kotlin/rxfirebase2/config/RxFirebaseRemoteConfig.kt
@file:Suppress("NOTHING_TO_INLINE")

package rxfirebase2.config

import com.google.firebase.remoteconfig.FirebaseRemoteConfig
import com.google.firebase.remoteconfig.FirebaseRemoteConfigSettings

import io.reactivex.Completable
import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.Single

inline fun FirebaseRemoteConfig.fetches()
        : Completable
        = RxFirebaseRemoteConfig.fetches(this)

inline fun FirebaseRemoteConfig.fetches(timeout: Long)
        : Completable
        = RxFirebaseRemoteConfig.fetches(this, timeout)

inline fun FirebaseRemoteConfig.fetchesDev()
        : Completable
        = RxFirebaseRemoteConfig.fetchesDev(this)
